lib/package.js

```diff
@@ -760,31 +761,50 @@ exports.createMain = function(pkg, pjson, downloadDir) {
     }
 
     if (main) {
-      if (main.startsWith('./'))
-        main = main.substr(2);
-      if (main.endsWith('.js'))
-        main = main.substr(0, main.length - 3);
+      if (main.startsWith('./')) {
+        main = main.substr(2);    
+      }  
+    }
+    else {
+      main = 'index';
     }
 
-    main = main || 'index';
-
-    // try the package.json main
     return new Promise(function(resolve) {
-      mainPath = path.resolve(downloadDir, main + '.js');
+      // Check to see if main points to an actual file or not
+      // i.e. does it include extension.
+      mainPath = path.resolve(downloadDir, main);
+      fs.exists(mainPath, resolve);
+    });
+
+  }).then(function(exists) {
+    if (exists) {
+      var match = /(\.[^.]+)$/.exec(main);
+      if (match) {
+        mainFileExtension = match[1];
+        main = main.slice(0, - mainFileExtension.length);
+      }
+      return exists;     
+    }
+
+    return new Promise(function(resolve) {      
+      mainPath = path.resolve(downloadDir, main + mainFileExtension);
```

Comment:
This is checking if `pjson.main` does not point to a file directly, meaning the Default extension needs to apply.